### DISCUSSION TITLE
Project layout and coding standards

### DISCUSSION THREAD

**Discussion by adam-huganir** [pypdfium2-team/pypdfium2#11] [General]
I have a way of organizing my company's projects previously that works very well in coordinating development across multiple devs, and I is pretty pep compliant. I can create a branch to show the layout and see if you are interested in adopting it. The basic tooling is:

-  `poetry` for virtualenvs, dependency management, and building/publishing
- `black`, `isort`, and `flake8` for consistent formatting
- `pytest` for testing
- `sphinx` for documentation
- `pre-commit` to ensure the formatting, testing, and documentation are enforced for contributors

Any or all of these can really benefit projects I think.

Thoughts?

**Comment by mara004** [comment]
Sounds good, but I'm not sure if we really need all of these tools. This kind of depends on how we want to define the scope of the project, as you said in the other thread.

So far I've only used Sphinx (and like it a lot), but the other tools I don't really know. I'm not experienced with maintaining software projects, to be honest. I'm just working on this out of personal need, and don't have the professional background that you have.

Currently time is a bit short for me and I may be slow to respond for the next time, but if you want to look into setting up these tools, I'm definitely okay with it :)

**Comment by adam-huganir** [comment]
@mara004 Can you set me up with write access you when get a chance? Can't fork or create a branch at the moment.

**Comment by mara004** [comment]
Sure. I've changed your role to Admin - hopefully it works now.

**Comment by mara004** [comment]
The forking problem still persists, but I'm afraid there's nothing I can do about this - the `Allow forking` option in the settings cannot be changed. Probably a limitation of private repositories on GitHub.

**Comment by adam-huganir** [comment]
I suspect so as well. I have a branch but I've had a crazy few weeks and haven't touched it since then. I need to do some reading on how bdist/wheels work, I've never had to work with them before. Hopefully I can put some time aside for it this weekend

**Comment by mara004** [comment]
Sure, take your time :)

**Comment by mara004** [comment]
By the way, I have enabled the GitHub discussions panel for our repository and started a first discussion.
And I have invited @YinlinHu, the original author of pypdfium, to join our organisation. Not sure why I forgot to do so earlier...

**Comment by YinlinHu** [comment]
@mara004 Thanks for having me. I may not have much time on this project, but happy to have discussions sometimes. :-)

**Comment by mara004** [comment]
Thank *you* for having initiated this project - I wouldn't have known at all how to do this without the initial PyPDFium, its examples and documentation. PyPDFium has been an invaluable help for my PDF tasks, and I'm very thankful that you have created it. You have done the main work for this project.
